### ISSUE TITLE
Replace PNG UI elements

### ISSUE THREAD

**Issue by SeeBeyondDev** [GDColon/GDBrowser#32]
I noticed that many of the images on the site are not actually used to their full resolution. There are many of them used as UI elements. This generally considered a bad practice but one that is hard to circumvent since many of GD's UI elements are shiny and complex graphics. I have some alternatives that can save on loading dozens of PNGs. 1. Using CSS for the more simple elements. An example of this is on the nav buttons for https://gdforums.com 2. Replacing PNGs with SVGs. GD's graphics aren't that complex and are generally made from many simple shapes and gradients. Converting the PNGs into SVGs, although time consuming due to some edge cases, can dramatically reduce the amount of data that gets sent on page load.

**Comment by siky112** [comment]
sorry but what is a svg

**Comment by SeeBeyondDev** [comment]
@siky112 An SVG (Scalable Vector Graphic) is an image format that instead of being made of pixels like a PNG (or any other type of raster graphics) is made out of plaintext. Here is an SVG file on my website. [Click here.](https://seebeyond.dev/assets/example.svg) you can view this in your browser as an image. But you can actually open up the file in a text editor and it is in XML format.
`<svg height="210" width="400">
  <path d="M150 0 L75 200 L225 200 Z" />
</svg>`
Since it is in this format it is much easier to edit and is very popular on the web.

**Comment by GDColon** [comment]
I've worked with vectors before, but I just didn't use them for GDBrowsers since I just ripped all the sprites directly from the game files. RobTop actually said that he makes all his GD stuff in vector but converts them to PNG afterwards and it hurts to read it every time. Anyways, I think it's a great idea but I'm mean and stubborn and personally feel like leaving it as PNGs for my own efficiency. I know I know, I suck 😛

**Comment by SeeBeyondDev** [comment]
If I were to make the changes would you consider accepting the pull request?

**Comment by GDColon** [comment]
I personally think it's not worth the trouble. Can you go over what exactly you would change? (and how)

**Comment by SeeBeyondDev** [comment]
I would first get the player shapes from /icons and turn them into multi-layered SVGs. This way you can save on the amount of files into 1 file per shape. For color changing I could use simple JS or a library such as [SVG.js](https://svgjs.com/docs/3.0/) to change the colors within the icon generator. It's very light and easier than manipulating PNGs. Second, I would convert as many UI elements from /assets as possible. I tested it and the conversion is worth it for the files. Even the small images for social media went down ~2kb in size. Manually re-creating them will take longer but will make the size even smaller than using an automatic SVG converter. In the site, the files will simply be replaced. The dimensions of the bitmap files will be included in the SVG file itself so there shouldn't be many problems with scaling.

**Comment by 101arrowz** [comment]
+1 I'm gonna side with SeeBeyondDev on this one. SVGs are indeed much better suited for this kind of thing (static icons), and there are tools to convert PNGs to SVGs. Either one of us could do it.

For icon generation, unfortunately, we can't actually use SVGs unless we completely redesign `icon.js`. Jimp does not support SVG output. We'd have to use something like node-potrace to trace the PNG into an SVG.

The best solution is to recolor on client side. If we use SVGs, it's actually much easier to do on client than on server side. Also, cropping is super simple, since SVGs are actually very similar to HTML elements.

**Comment by GDColon** [comment]
Buzzkill Punctuation here - I'd rather not change icon.js since I really want them to be PNG images, and also want the icons ripped directly from the game files.

**Comment by 101arrowz** [comment]
Any reason in particular that you want them to be PNG? Rendering an SVG to a PNG is much less computationally expensive than the reverse, so if you wanted to offer a PNG option in the API, we could do that while still creating the images in SVG originally. The PNG option would just have the server render the generated SVG to a PNG of size myWidth by myHeight.

Also, if you wanted them _directly_ from the game files, you shouldn't have merged my PR with speed optimizations. I shrunk the file sizes of all the PNGs by compressing them. This is really similar, it's just a different format of holding the image data.

I actually feel quite strongly about implementing this one. It's a very good idea and would solve a lot of your server lag issues. Plus, it renders better across devices because it's a vector. PLUS, I can make a tool to automatically convert any PNGs into SVG, so if any new icons are added we don't need to do any manual work.

**Comment by GDColon** [comment]
Main reason is that it's easier on the client. It's much better if they can save the icons as PNGs than vectors that can't be edited or pasted into programs. I just feel like this is overall too fancy of a change for my liking.

**Comment by 101arrowz** [comment]
Who said SVG files can't be edited? More than that, why do users need to copy images from GDBrowser? If they need an SVG, they can use the hypothetical PNG option when calling server side.

**Comment by SeeBeyondDev** [comment]
TL;DR
- Image processing on the server can be a bottleneck on performance
- Images can be **processed as SVGs** then exported and **be saved PNGs** for convenience if thats what you prefer.
- Optionally if needed we can have advanced options for exporting such as filetype or size

What operations are handled by the server? Perusing the package.json I see you are using jimp. Why is the server handling images? This is a major disadvantage in the event of high traffic the server is overloaded. Ideally, in this type of application intensive processes like this should be handled by the client. May I also ask where this node server is being hosted? I hadn't noticed before but this can be a large bottleneck to performance. I think it would not only be beneficial to 1. Move as many intensive processes such as image processing off the server load as these tasks can impede the server when it is trying to complete requests. 2. And Arrowz, are you suggesting that images be exported as SVG by default? Because on that I actually disagree as exporting as a PNG would be more convenient. We can for sure have imaged be exported as a PNG as conversion from SVG to PNG is trivial.

**Comment by GDColon** [comment]
I'm starting to get really lost here. I'm still not sure about this change, probably because I don't like code I don't know, and because I just don't like change. I'm sure it's good. If you were to make a PR with all the changes you brought up, can you explain what it would look like? (pseudocode would help)

Really sorry if I'm being super annoying/stubborn/infuriating. I am very aware of this issue and am looking for a fix. I just really like having things exactly how I want them, I guess. ~~also all your opinions are wrong~~

**Comment by 101arrowz** [comment]
Default options are trivial to change. So sure, the default behavior can be PNG, and SVG can be an option. That SVG option will save some server resources and be used by default by GDBrowser if these ideas are added.

@GDColon I'll work on this SVG stuff and submit a PR. If you really REALLY hate this idea then let me know, and I'll submit a PR using [sharp](https://npmjs.com/package/sharp), which is 30x faster than Jimp. I've used it, it's super easy to work with.

<strike>also you should learn new things instead of trying to make everything work with only what you know, boomer</strike>
<strike>also jquery is shit and i'm still salty that you didn't accept my jqueryless pr</strike>

**Comment by GDColon** [comment]
hahahaa

but yeah you might wanna go over what exactly you're doing in the pr before i just shut it down like the monster/boomer i am

**Comment by 101arrowz** [comment]
Yeah let's do more of this conversation stuff before I spend *cough* 5 DAYS WORKING ON SOMETHING FOR IT TO GET REJECTED AAAAAA *cough*.

So this will entail removing all Jimp references and replacing them with SVG manipulation. This could be done with something like [simple-svg-tools](https://www.npmjs.com/package/simple-svg-tools). I'm not sure of the specifics but I can figure it out. Alternatively, @SeeBeyondDev could do it.

As for the `sharp` idea, it's very easy to change. I could do that alternatively. Your call.

**Comment by SeeBeyondDev** [comment]
@101arrowz nonono lets not involve nodejs in this. SVG manipulation can be done just fine on the client side as well. One of the main points of SVG switch was that it was an easier format to edit since it's XML and it could be done on the client side. Also, colon I'd be glad to explain my proposed changes just give me some time on that pseudocode PR mockup since I have to see some of the places I would need to modify.

**Comment by GDColon** [comment]
h

**Comment by 101arrowz** [comment]
Colon, your speech is so moving and inspiring. I could never [h](https://www.youtube.com/watch?v=DwjbP7aihxQ)ave said something as meaningful and incisive as you.

@SeeBeyondDev I agree with that concept, but wouldn't that mean we would have to use a separate API for GDBrowser client side and for the general public to use? That seems complicated.

**Comment by GDColon** [comment]
lmao that's one of my favorite videos on youtube, good taste

**Comment by 101arrowz** [comment]
So, beyond memery, what do you think of SeeBeyondDev's idea?

**Comment by GDColon** [comment]
Honestly, you two lost me a while ago

**Comment by 101arrowz** [comment]
:crab: cologne's sanity is GONE :crab:

I'll make the PR with sharp so we have a quick fix. I don't expect that SeeBeyondDev will be able to understand the spaghetti quickly, so I'll just do it now (only somewhat kidding).

**Comment by SeeBeyondDev** [comment]
@101arrowz What do you mean separate? We just export SVG as a PNG. We don't need to manipulate the PNG, it's just for exporting. And what do you mean the general public? What is the opposite of that? Why do you want the server to be used for editing images?

**Comment by 101arrowz** [comment]
Here's the deal. If you want Node.js to recolor the SVGs, which is easy anyway, then we can render it to a PNG and send to client. If the client requests an SVG, we send the SVG without rendering.

If you want the client to recolor the SVGs, how do we continue exporting as PNG? We need color data in the SVG for that. We'd have to have completely separate pipelines if the request is made from GDBrowser or from a developer who's calling the API, expecting a PNG. We'd need to continue the bad Jimp system for the normal API caller, and have the better system return an uncolored SVG along with color data to GDBrowser. It's overly complicated. That's why SVG recoloring on server side is optimal.

**Comment by SeeBeyondDev** [comment]
OOOH I understand where the confusion was! Okay yeah I get where you are coming from. Yeah returning a PNG is good for the api. but perhaps we can have a flag for the api that returns an svg when on the client so the client can have color changes be immediate instead of having to wait for the server to do it. So a regular api user can keep their PNGs but since we can offload work onto people accessing the actual webpage we can load an asset once and recolor there. I hope that’s clearer.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by GDColon** [comment]
I have no idea what the hell this chat is anymore so here's the one TRUE vector

![image](https://user-images.githubusercontent.com/56617416/68536904-6cf82b80-0328-11ea-89d9-b0aaecd7476e.png)

**Comment by 101arrowz** [comment]
colon you are truly evil, that image should have a trigger warning, my second-to-last braincell died watching that movie and my last one is about to disintegrate seeing that character do the T pose

So I'm gonna leave resolving this issue to @SeeBeyondDev, since I'm not sure how to implement this system in a clean way. Thanks for contributing to the project!

**Comment by GDColon** [comment]
Come on man, the first movie wasn't that bad. They just started to capitalize on the minions after that.
_Vector! That's me! 'cuz I'm commiting crimes with both **direction** and **magnitude**! Oh yeah!_

**Comment by jbmagination** [comment]
this is the best t*h*read. but really, *vector*s would be great
